### PR TITLE
Sql optimise new index merged branch

### PR DESCRIPTION
**Pull request by nadeeshaan**
## Testing Instructions:

This PR consists all the PRs merged in to this one regarding the index additions. Two approaches can be followed in order to test this PR. One is in an already installed site and one is for fresh distributions.
#### For already installed sites

For already installed sites first apply the PR and then add follow the bellow commands in order to add the indexes manually to the database
- Index added to #_languages table (_You can see this query execution when you access the site_)

`ALTER TABLE #__languages ADD INDEX 'idx_published_ordering' ('published','ordering');` (#11)

**Related Query Before changes**

``` sql
  SELECT *
  FROM #__languages
  WHERE published=1
  ORDER BY ordering ASC
```

**Related Query Before changes**

``` sql
  SELECT lang_id,lang_code,title,title_native,sef,image,description,
                 metakey,metadesc,sitename,published,access,ordering
  FROM #__languages
  WHERE published=1
  ORDER BY ordering ASC
```

---
- Index added to #_session table (_Query executes every time a user access a page_)

`ALTER TABLE #__session DROP PRIMARY KEY, ADD PRIMARY KEY(session_id(32))` ( #8 )
`ALTER TABLE #__session DROP INDEX time, ADD INDEX time(time(10))`

**Index Description**
This Index added in order to speed up the session update query allowing the sub string index usages

---
- Index added to #_template_styles table (_You can see this query execution when you access the site_)

`ALTER TABLE #__template_styles ADD INDEX idx_client_id('client_id')`  ( #13 )

**Index Description**
Without applying this index the below query cannot use any index to retrieve the rows from the database. After applying this index it clearly shows in the explanation section that the query uses the index to retrieve the rows

**Related Query**

``` sql
    SELECT id, home, template, s.params 
    FROM #__template_styles as s 
    LEFT JOIN #__extensions as e 
    ON e.element=s.template 
    AND e.type='template' 
    AND e.client_id=s.client_id 
    WHERE s.client_id = 0 AND e.enabled = 1
```

---
- Index added to #__contentitem_tag_map table (_You can see this query execution when you access the site_)

`ALTER TABLE #__contentitem_tag_map ADD INDEX idx_alias_item_id ('type_alias','content_item_id')`  ( #14 )

**Index Description**
Without applying this index the below query cannot use any index to retrieve the rows from the database. After applying this index it clearly shows in the explanation section that the query uses the index to retrieve the rows. Also one another change did relevant to this index changes. Which is removing the `t.*` from the relevant query and adding the table fields instead. This change can be found (#14) and please refer to that PR also.

 **Related Query Before changes**

``` sql
    SELECT m.tag_id,t.*
    FROM #__contentitem_tag_map AS m
    INNER JOIN #__tags AS t ON m.tag_id = t.id
    WHERE m.type_alias = 'com_content.article' AND m.content_item_id = 5196 AND t.published = 1
    AND t.access IN (1,1,5)
```

**Related Query After changes**

``` sql
    SELECT m.tag_id,'id',t.parent_id,t.lft,t.rgt,t.level,t.path,t.title,t.alias,t.note,t.description,t.published,
    t.checked_out,t.checked_out_time,t.access,t.params,t.metadesc,t.metakey,t.metadata,
    t.created_user_id,t.created_time,t.created_by_alias,t.modified_user_id,t.modified_time,
    t.images,t.urls,t.hits,t.language,t.version,t.publish_up,t.publish_down
    FROM #__contentitem_tag_map AS m
    INNER JOIN #__tags AS t ON m.tag_id = t.id
    WHERE m.type_alias = 'com_content.article' AND m.content_item_id = 5196 AND t.published = 1
    AND t.access IN (1,1,5)
```

---
- Index added to #_extensions table (_You can see this query execution when you access the site_)

`ALTER TABLE #__extensions ADD INDEX 'idx_type_ordering' ('type','ordering')` ( #16 )

**Index Description**
Without applying this index the below query uses `filesort` to retrieve the rows from the database, which is an expensive retrieval operation. After applying this index it clearly shows in the explanation section that the query uses the index to retrieve the rows except the `filesort`.

**Related Query**

``` sql
    SELECT folder AS type, element AS name, params
    FROM #_extensions
    WHERE enabled >= 1
    AND type ='plugin'
    AND state >= 0
    AND access IN (1,1,5)
    ORDER BY ordering
```

---
- Index added to #_menu table (_You can see this query execution when you access the site_)

`ALTER TABLE #__menu ADD INDEX 'idx_client_id_published_lft' ('client_id','published','lft')` ( #17 )

**Index Description**
Without applying this index the below query uses `filesort` to retrieve the rows from the database, which is an expensive retrieval operation. After applying this index it clearly shows in the explanation section that the query uses the index to retrieve the rows except the `filesort`.

**Related Query**

``` sql
    FROM #_menu AS m
    LEFT JOIN #_extensions AS e
    ON m.component_id = e.extension_id
    WHERE m.published = 1
    AND m.parent_id > 0
    AND m.client_id = 0
    ORDER BY m.lft
```

---
- Index added to #__viewlevels table
  (_query Executes when clicking on the add new category in the admin console_)

Add New Category

`ALTER TABLE #__viewlevels 
            ADD INDEX`idx_ordering_title`(`ordering`,`title`)`(  #21 )

**Index Description**
Without applying this index the below query uses `filesort` to retrieve the rows from the database, which is an expensive retrieval operation. After applying this index it clearly shows in the explanation section that the query uses the index to retrieve the rows except the `filesort`. Also changed the `JROOT/libraries/cms/html/access.php` file's query which is shown below by removing the `GROUP BY` clause since the `PRIMARY KEY` includes in the selected fields. And this reduce the execution time

**Related Query Before Change**

``` sql
    SELECT a.id AS value, a.title AS text
    FROM ltzvy_viewlevels AS a
    GROUP BY a.id, a.title, a.ordering
    ORDER BY a.ordering ASC,`title` ASC
```

**Related Query After Change**

``` sql
    SELECT a.id AS value, a.title AS text
    FROM ltzvy_viewlevels AS a
    ORDER BY a.ordering ASC,`title` ASC
```

### CODE REVIEW

**Comment by waader** [comment]
Can you add the indexes for postresql and mssql too?

**Comment by zero-24** [comment]
@nadeeshaan 

Github say that
`This pull request contains merge conflicts that must be resolved.`

Can you have a look into it?

File: installation/sql/mysql/joomla.sql [hunk 1]
```diff
@@ -443,7 +443,8 @@ CREATE TABLE IF NOT EXISTS `#__contentitem_tag_map` (
   KEY `idx_date_id` (`tag_date`,`tag_id`),
   KEY `idx_tag` (`tag_id`),
   KEY `idx_type` (`type_id`),
-  KEY `idx_core_content_id` (`core_content_id`)
+  KEY `idx_core_content_id` (`core_content_id`),
+	KEY 'idx_alias_item_id' ('type_alias','content_item_id')
```
Comment (gunjanpatel) [inline installation/sql/mysql/joomla.sql]: Thanks @Sophist-UK I will fix this issues.

**Comment by gunjanpatel** [comment]
@zero-24 thanks for review. I am working on resolving conflict right now.

**Comment by gunjanpatel** [comment]
This PR is conflict is now solved and ready to review. Thanks everyone for review and tests.

**Comment by wnnz34** [comment]
the Query 
SELECT `m`.`core_content_id`,`m`.`content_item_id`,`m`.`type_alias`,COUNT( `tag_id`) AS `count`,`ct`.`router`,`cc`.`core_title`,`cc`.`core_alias`,`cc`.`core_catid`,`cc`.`core_language`,`cc`.`core_params`

  FROM `sbgcn_contentitem_tag_map` AS `m`

  INNER JOIN `sbgcn_tags` AS `t` 
  ON m.tag_id = t.id

  INNER JOIN `sbgcn_ucm_content` AS `cc` 
  ON m.core_content_id = cc.core_content_id

  INNER JOIN `sbgcn_content_types` AS `ct` 
  ON m.type_alias = ct.type_alias

  WHERE `m`.`tag_id` IN (2,3,4,5) 
  AND t.access IN (1,1) 
  AND (cc.core_access IN (1,1) OR cc.core_access = 0) 
  AND (`m`.`content_item_id` <> 24 OR `m`.`type_alias` <> 'com_content.article') 
  AND `cc`.`core_state` = 1  
  AND (`cc`.`core_publish_up`='0000-00-00 00:00:00' OR `cc`.`core_publish_up`<='2015-08-21 11:02:01') 
  AND (`cc`.`core_publish_down`='0000-00-00 00:00:00' OR `cc`.`core_publish_down`>='2015-08-21 11:02:01')

  GROUP BY `m`.`core_content_id`,`m`.`content_item_id`,`m`.`type_alias`,`ct`.`router`,`cc`.`core_title`,`cc`.`core_alias`,`cc`.`core_catid`,`cc`.`core_language`,`cc`.`core_params`

  ORDER BY `count` DESC 
  LIMIT 0, 5
Tested has a Query Time of 15.00 ms. and it is slower than some MSQL Queries.

SELECT folder AS type, element AS name, params
    FROM #_extensions
    WHERE enabled >= 1
    AND type ='plugin'
    AND state >= 0
    AND access IN (1,1,5)
    ORDER BY ordering
tested ! and all the others MYSQL QUERIES.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/4115">issues.joomla.org/joomla-cms/4115</a>.</sub>

**Comment by euismod2336** [comment]
I have tested this item :red_circle: unsuccessfully on 7e0ed08c30def298adb3c6f62befb4845dc5a94e<br /><br />Confirmed for all but #17, see screenshot. I assumed that also here the filesort shouldn't be used anymore
Screenshot<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/4115">issues.joomla.org/joomla-cms/4115</a>.</sub>

**Comment by beat** [comment]
@euismod2336 That's not necessarily an unsuccessful test! Specially if master has no better result... Many other improvements, worthwhile ones were made here.  :-)

sure an adequate index seems to exist: `(client_id, published, lft)` , and would be most probably used by MySQL depending on the size of the result. The issue here for not being able to use straight that index is the `m.parent > 0` in the where statement. You should try again with maybe 1000 or 10000 rows to see if MySQL uses that index.

**Comment by wilsonge** [comment]
No but this PR does conflict because it predates all the utf8mb4 index changes - there's no point in testing this unless the branch is updated again

**Comment by wilsonge** [comment]
For when @brianteeman hits this in a few weeks at the sprint tho - this is one that needs redoing - not just closing. It's going to give big improvements for "free"

**Comment by euismod2336** [comment]
@beat i agree that the improvements are considerable, but not all issues were "fixed" (as far as i could see), i'll have another go at it with more records if it might change it. all the other queries showed direct results.

@wilsonge not sure about the utf8mb4 changes, but this was PR was tested on the staging branch.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/4115">issues.joomla.org/joomla-cms/4115</a>.</sub>

**Comment by euismod2336** [comment]
I have tested this item :white_check_mark: successfully on 7e0ed08c30def298adb3c6f62befb4845dc5a94e<br /><br />After adding 10.000 records the query is done properly on the index.

Thank you @beat for pointing this out.

Image
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/4115">issues.joomla.org/joomla-cms/4115</a>.</sub>

**Comment by zero-24** [comment]
Set to "closed" on behalf of @zero-24 by The <a href="https://github.com/joomla/jissues">JTracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/4115">issues.joomla.org/joomla-cms/4115</a>

**Comment by zero-24** [comment]
Closing as we have a new updated PR at #10284 by @alikon <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/4115">issues.joomla.org/joomla-cms/4115</a>.</sub>

**Pull request closed without merging**